lxd/scriptlet/starlark_test.go

```diff
@@ -0,0 +1,189 @@
+package scriptlet_test
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/lxc/lxd/lxd/scriptlet"
+	"go.starlark.net/starlark"
+)
+
+type starlarkMarshalTest struct {
+	from   any
+	to     starlark.Value
+	expect string
+}
+
+func (s *starlarkMarshalTest) run(t *testing.T) {
+	defer func() {
+		if err := recover(); err != nil {
+			t.Errorf("Unexpected panic: %v", err)
+		}
+	}()
+
+	name := fmt.Sprintf("%#v (%T)", s.from, s.from)
+	converted, err := scriptlet.StarlarkMarshal(s.from)
+
+	if err == nil && s.expect != "" {
+		t.Errorf("Expected error converting %v", name)
+	} else if err != nil && (s.expect == "" || !strings.HasPrefix(err.Error(), s.expect)) {
+		t.Errorf("Unexpected error converting %v: %v", name, err)
+	}
+
+	if !reflect.DeepEqual(s.to, converted) {
+		t.Errorf("Incorrect value returned converting %v: expected %v but got %v", name, s.to, converted)
+	}
+}
+
+type DummyStringer int
+
+var _ fmt.Stringer = DummyStringer(0)
+
+func (DummyStringer) String() string { return "(DummyStringer)" }
+
+func TestStarlarkMarshal(t *testing.T) {
+	type DummyEmbeddedStruct struct {
+		A string
+	}
+
+	tests := []starlarkMarshalTest{{
+		from: starlark.MakeInt(-1),
+		to:   starlark.MakeInt(-1),
+	}, {
+		from: nil,
+		to:   starlark.None,
+	}, {
+		from: true,
+		to:   starlark.True,
+	}, {
+		from: int(-1),
+		to:   starlark.MakeInt(-1),
+	}, {
+		from: int8(-1),
+		to:   starlark.MakeInt(-1),
+	}, {
+		from: int16(-1),
+		to:   starlark.MakeInt(-1),
+	}, {
```

Comment:
Why is int32 check missing here?

Reply:
The `int32` check is missing here as `int32` and `rune` are indistinguishable (`rune` is just a type alias) and as such `StarlarkMarshal` cannot support both. As such, the choice was between supporting a fifth signed-integer type or something more interesting. I figured it would be more intuitive for someone to pass a rune to `MarshalStarlark` and for a length-1 string to appear rather than an integer representing its UTF-8 value.

I'm open to changing this though.

Reply:
Makes sense. As discussed on MM lets stick with handling int32 for now, and document that runes will be treated the same way/not supported.